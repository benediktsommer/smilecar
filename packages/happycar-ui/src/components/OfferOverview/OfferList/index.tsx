import { memo } from 'react';
import { Box, Card, CardContent, withStyles } from '@material-ui/core';

import { useGetFilteredOffers } from '../../../hooks/useGetFilteredOffers';

import { Offer } from './Offer';
import { OfferListView } from './OfferListView';

export const OfferList = memo(() => {
  const { offers } = useGetFilteredOffers();

  return <OfferListView offers={offers} />;
});
