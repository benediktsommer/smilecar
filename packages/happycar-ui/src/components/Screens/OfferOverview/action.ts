import { IOffer } from '../../../interfaces/offer.interface';
import { IFilter } from '../../../interfaces/filter.interface';

export enum OfferListActionTypes {
  OFFERLIST_LOAD_START = 'OFFERLIST_LOAD_START',
  OFFERLIST_LOAD_SUCCESSFUL = 'OFFERLIST_LOAD_SUCCESSFUL',
  OFFERLIST_LOAD_FAILED = 'OFFERLIST_LOAD_FAILED',
  OFFER_FILTERED = 'OFFER_FILTERED',
}

interface IOfferListLoadStart {
  type: OfferListActionTypes.OFFERLIST_LOAD_START;
}

interface IOfferListLoadSuccessful {
  type: OfferListActionTypes.OFFERLIST_LOAD_SUCCESSFUL;
  offers: IOffer[];
  filter: IFilter;
}

interface IOfferListLoadFailed {
  type: OfferListActionTypes.OFFERLIST_LOAD_FAILED;
  error: string;
}

interface IOfferListFiltered {
  type: OfferListActionTypes.OFFER_FILTERED;
  filterIdentifier: keyof IFilter;
  filterOption: string | number;
}

export type OfferListStateActions =
  | IOfferListLoadStart
  | IOfferListLoadSuccessful
  | IOfferListLoadFailed
  | IOfferListFiltered;
