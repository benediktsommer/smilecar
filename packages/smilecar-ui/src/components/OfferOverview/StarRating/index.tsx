import { memo } from 'react';
import { Box, Typography, withStyles } from '@material-ui/core';

import { IOffer } from '../../../interfaces/offer.interface';

interface IProps {
  rating: IOffer['supplier']['rating'];
}

const StarRatingOverall = withStyles((theme) => ({
  root: {
    background: 'url(./grey.png) no-repeat',
    height: '16px',
    maxWidth: '100px',
  },
}))(Box);

const StarRatingOverallFill = withStyles((theme) => ({
  root: {
    background: 'url(./yellow.png) no-repeat',
    height: '16px',
    maxWidth: '100px',
  },
}))(Box);

export const StarRating = memo(({ rating }: IProps) => (
  <Box>
    <StarRatingOverall>
      <StarRatingOverallFill style={{ width: `${rating.average * 100}%` }} />
      <Typography variant="subtitle2" component="span">
        &#216; {Math.round(5 * rating.average * 100) / 100} ({rating.count})
      </Typography>
    </StarRatingOverall>
  </Box>
));
