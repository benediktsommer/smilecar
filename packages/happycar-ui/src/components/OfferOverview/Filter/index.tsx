import { Fragment, memo, ReactNode, useCallback } from 'react';
import { Box } from '@material-ui/core';

import { FilterTypes } from '../../../interfaces/filter.interface';
import { useGetFilter } from '../../../hooks/useGetFilter';
import { useOfferListDispatch } from '../../Screens/OfferOverview/context';
import { OfferListActionTypes } from '../../Screens/OfferOverview/action';

import { FilterWidget } from './FilterWidget';

export const Filter = memo(() => {
  const { dispatch } = useOfferListDispatch();
  const { filter } = useGetFilter();

  const handleFilter = useCallback(
    (filterType: FilterTypes, value: number | string): void => {
      dispatch({
        type: OfferListActionTypes.OFFER_FILTERED,
        filterIdentifier: filterType,
        filterOption: value,
      });
    },
    [dispatch]
  );

  if (!filter) {
    return null;
  }

  return (
    <Box mt={4} mr={2}>
      {Object.keys(filter).map((option) => {
        /* some strage TS behaviors */
        const filterOption = filter[option as FilterTypes];
        if (!filterOption) {
          return null;
        }
        /* end strange behavior */

        return (
          <Fragment key={option}>
            <FilterWidget
              headline={option.toUpperCase()}
              handleFilter={handleFilter}
              filterOptions={filterOption.map((value) => ({
                identifier: option as FilterTypes,
                disabled: false,
                value,
              }))}
            />
          </Fragment>
        );
      })}
    </Box>
  );
});
