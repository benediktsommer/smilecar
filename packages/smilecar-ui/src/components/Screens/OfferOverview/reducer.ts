import { IOffer } from '../../../interfaces/offer.interface';
import { IFilter } from '../../../interfaces/filter.interface';

import { OfferListActionTypes, OfferListStateActions } from './action';

interface ILoadingOfferState {
  isLoading: true;
  error: null;
  offers: null;
  filter: null;
}

interface IErroredOfferState {
  isLoading: false;
  error: string;
  offers: null;
  filter: null;
}

interface ILoadedOfferState {
  isLoading: false;
  error: null;
  offers: IOffer[];
  filter: IFilter;
}

export type IOfferState =
  | ILoadingOfferState
  | IErroredOfferState
  | ILoadedOfferState;

export const initialOfferState: IOfferState = {
  error: null,
  isLoading: true,
  offers: null,
  filter: null,
};

export const offerReducer = (
  state: IOfferState,
  action: OfferListStateActions
): IOfferState => {
  switch (action.type) {
    case OfferListActionTypes.OFFERLIST_LOAD_SUCCESSFUL: {
      return {
        ...state,
        error: null,
        isLoading: false,
        offers: action.offers.map((offer) => ({
          ...offer,
          isActive: true,
        })),
        filter: action.filter,
      };
    }

    case OfferListActionTypes.OFFERLIST_LOAD_FAILED: {
      return {
        ...state,
        isLoading: false,
        error: action.error,
        offers: null,
        filter: null,
      };
    }

    case OfferListActionTypes.OFFERLIST_LOAD_START: {
      return {
        ...state,
        isLoading: true,
        error: null,
        offers: null,
        filter: null,
      };
    }

    case OfferListActionTypes.OFFER_FILTERED: {
      if (!state.offers) {
        return state;
      }

      return {
        ...state,
        offers: state.offers.map((offer) => {
          if (
            offer.vehicle[action.filterIdentifier] &&
            offer.vehicle[action.filterIdentifier] === action.filterOption
          ) {
            return {
              ...offer,
              isActive: true,
            };
          }

          return {
            ...offer,
            isActive: false,
          };
        }),
      };
    }

    default: {
      return state;
    }
  }
};
