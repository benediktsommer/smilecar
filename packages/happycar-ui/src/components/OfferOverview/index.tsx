import { memo } from 'react';
import { Grid } from '@material-ui/core';

import { useLoadOffers } from '../../hooks/useLoadOffers';

import { Filter } from './Filter';
import { OfferList } from './OfferList';
import { FilterProvider } from './Filter/context';

export const OfferOverview = memo(() => {
  useLoadOffers();

  return (
    <FilterProvider>
      <Grid container>
        <Grid item xs={12}></Grid>
        <Grid item xs={12} md={3}>
          <Filter />
        </Grid>
        <Grid item xs={12} md={9}>
          <OfferList />
        </Grid>
      </Grid>
    </FilterProvider>
  );
});
