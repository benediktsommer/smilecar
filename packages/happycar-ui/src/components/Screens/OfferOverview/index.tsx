import { memo } from 'react';

import { OfferOverview } from '../../OfferOverview';

import { OfferProvider } from './context';

export default memo(() => {
  return (
    <OfferProvider>
      <OfferOverview />
    </OfferProvider>
  );
});
