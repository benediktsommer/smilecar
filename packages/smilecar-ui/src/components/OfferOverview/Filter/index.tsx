import { Fragment, memo, ReactNode, useCallback } from 'react';
import { Box, Button, withStyles } from '@material-ui/core';

import { FilterTypes } from '../../../interfaces/filter.interface';
import { useGetFilter } from '../../../hooks/useGetFilter';
import { useOfferListDispatch } from '../../Screens/OfferOverview/context';
import { OfferListActionTypes } from '../../Screens/OfferOverview/action';
import { useGetAvailableFilter } from '../../../hooks/useGetAvailableFilter';

import { FilterWidget } from './FilterWidget';
import { useFilterState } from './context';

const FilterWrapper = withStyles((theme) => ({
  root: {
    [theme.breakpoints.up('md')]: {
      padding: theme.spacing(0, 2, 0, 0),
    },
  },
}))(Box);

const ResetFilterCTA = withStyles((theme) => ({
  root: {
    [theme.breakpoints.down('md')]: {
      margin: theme.spacing(2, 0),
    },
  },
}))(Button);

export const Filter = memo(() => {
  const { addFilter, deleteFilter, filter } = useFilterState();
  const { filter: originalFilter } = useGetFilter();
  const { availableFilter } = useGetAvailableFilter();

  if (!originalFilter || !availableFilter) {
    return null;
  }

  return (
    <FilterWrapper>
      {Object.keys(originalFilter).map((option) => {
        const filterOption = originalFilter[option as FilterTypes] || [];
        if (!filterOption) {
          return null;
        }

        return (
          <Fragment key={option}>
            <FilterWidget
              headline={option.toUpperCase()}
              handleFilter={addFilter}
              filterOptions={filterOption.map((value) => ({
                identifier: option as FilterTypes,
                disabled: !availableFilter[option as FilterTypes]?.includes(
                  value
                ),
                active:
                  filter.selected[option as FilterTypes]?.includes(value) ||
                  false,
                value,
              }))}
            />
          </Fragment>
        );
      })}
      <ResetFilterCTA
        variant="contained"
        color="primary"
        size="large"
        type="submit"
        onClick={deleteFilter}
      >
        Filter zur√ºcksetzen
      </ResetFilterCTA>
    </FilterWrapper>
  );
});
