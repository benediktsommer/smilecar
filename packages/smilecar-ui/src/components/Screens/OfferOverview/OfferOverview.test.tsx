import { render } from '@testing-library/react';

import OfferOverview from './';

jest.mock('../../OfferOverview', () => ({
  OfferOverview: () => <div id="OfferOverview" />,
}));

describe('Test <OfferOverview />', () => {
  it('should render an html markup', () => {
    const { asFragment } = render(<OfferOverview />);
    expect(asFragment()).toMatchSnapshot();
  });
});
