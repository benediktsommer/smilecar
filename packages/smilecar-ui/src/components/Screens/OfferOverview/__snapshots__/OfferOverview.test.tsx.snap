// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <OfferOverview /> should render an html markup 1`] = `
<DocumentFragment>
  <div
    id="OfferOverview"
  />
</DocumentFragment>
`;
